SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/SteamMatchmaking.cs

```diff
@@ -0,0 +1,392 @@
+                body.num_members,
+                memberList,
+                existingLobby?.Distance,
+                existingLobby?.Weight
+            );
+
+            appLobbies[ lobbyId ] = updatedLobby;
+
+            Client.PostCallback( new LobbyDataCallback( body.app_id, updatedLobby ) );
+        }
+
+        void HandleUserJoinedLobby( IPacketMsg packetMsg )
+        {
+            var userJoinedLobby = new ClientMsgProtobuf<CMsgClientMMSUserJoinedLobby>( packetMsg );
+            CMsgClientMMSUserJoinedLobby body = userJoinedLobby.Body;
+
+            var lobbyId = new SteamID( body.steam_id_lobby );
+            var userId = new SteamID( body.steam_id_user );
+
+            var appLobbies = lobbies.GetOrAdd( body.app_id, k => new ConcurrentDictionary<SteamID, Lobby>() );
+            var lobby = appLobbies.ContainsKey( lobbyId ) ? appLobbies[ lobbyId ] : null;
+
+            if ( lobby == null )
+            {
+                // Unknown lobby
+                return;
+            }
+
+            int memberIndex = lobby.members.FindIndex( m => m.SteamID == userId );
+
+            if ( memberIndex >= 0 )
+            {
+                // Already in lobby
+                return;
+            }
+
+            var member = new Lobby.Member( new SteamID( body.steam_id_user ), body.persona_name );
+
+            lobby = lobby.Clone(); // Cloning to avoid mutating the original Lobby, as it was previously returned in a callback (not thread-safe).
+            lobby.members.Add( member );
+            appLobbies[ lobbyId ] = lobby;
+
+            Client.PostCallback( new UserJoinedLobbyCallback(
+                body.app_id,
+                new SteamID( body.steam_id_lobby ),
+                member
+            ) );
+        }
+
+        void HandleUserLeftLobby( IPacketMsg packetMsg )
+        {
+            var userLeftLobby = new ClientMsgProtobuf<CMsgClientMMSUserLeftLobby>( packetMsg );
+            CMsgClientMMSUserLeftLobby body = userLeftLobby.Body;
+
+            var lobbyId = new SteamID( body.steam_id_lobby );
+            var userId = new SteamID( body.steam_id_user );
+
+            var appLobbies = lobbies.GetOrAdd( body.app_id, k => new ConcurrentDictionary<SteamID, Lobby>() );
+            var lobby = appLobbies.ContainsKey( lobbyId ) ? appLobbies[ lobbyId ] : null;
+
+            if ( lobby == null )
+            {
+                // Unknown lobby
+                return;
+            }
+
+            var memberIndex = lobby.members.FindIndex( m => m.SteamID == userId );
+
+            if ( memberIndex < 0 )
+            {
+                // Not in lobby
+                return;
+            }
+
+            var member = lobby.members[ memberIndex ];
+
+            lobby = lobby.Clone(); // Cloning to avoid mutating the original Lobby, as it was previously returned in a callback (not thread-safe).
+            lobby.members.RemoveAt( memberIndex );
+            appLobbies[ lobbyId ] = lobby;
```

Comment:
Should we maybe return clones in the callback rather than having this racy clone-and-replace dance internally every time?

Reply:
I've refactored this based on earlier feedback. There is no longer a concept of clone. This logic has moved into a `LobbyCache` in a vaguely similar fashion to the `FriendCache`. Will push soon.